aptos-move/e2e-move-tests/src/tests/enum_upgrade_fv.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Tests for enum type upgrade compatibility
+
+use crate::{assert_success, assert_vm_status, tests::common, MoveHarness};
+use aptos_framework::BuildOptions;
+use aptos_language_e2e_tests::account::Account;
+use aptos_package_builder::PackageBuilder;
+use aptos_types::{account_address::AccountAddress, transaction::TransactionStatus};
+use move_core_types::vm_status::StatusCode;
+
+#[test]
+fn enum_upgrade() {
+    let mut h = MoveHarness::new();
+    let acc = h.new_account_at(AccountAddress::from_hex_literal("0x815").unwrap());
+
+    // Initial publish
+    let result = publish(
+        &mut h,
+        &acc,
+        r#"
+        module 0x815::m {
+            enum Data<T> has key {
+               V1{x: ||T has copy + store + drop }
+            }
+        }
+    "#,
+    );
+    assert_success!(result);
+
+    // incompatible variant
+    let result = publish(
+        &mut h,
+        &acc,
+        r#"
+        module 0x815::m {
+            enum Data<T> has key {
+               V1 {x: ||T has store}
+            }
+        }
+    "#,
+    );
+    assert_vm_status!(result, StatusCode::BACKWARD_INCOMPATIBLE_MODULE_UPDATE);
+
+    // identical variant
```

Comment:
Could we show an upgrade with increasing abilities as well (e.g., you could start off with abilities `x: || T has copy + store`, and try to upgrade with `x: || T has store` (which you already have), and then try to upgrade with `x: || T has  copy + store + drop`)?

Reply:
done